File: ports/fsharp/Push.fs
module SpriteTool.Push

open System
open System.Diagnostics
open System.IO

let private usage () =
    printfn
        "Usage: sprite-tool push <local-path> <remote-path> [sprite-name]\n\
         \n\
         Examples:\n\
         \  sprite-tool push ./file.txt /home/sprite/file.txt\n\
         \  sprite-tool push ./mydir /home/sprite/mydir\n\
         \  sprite-tool push ./file.txt /tmp/file.txt ember-red-hawk"

    exit 1

let run (args: string list) =
    if args.Length < 2 then
        usage ()

    let localPath = args.[0]
    let remotePath = args.[1]

    let spriteName =
        if args.Length > 2 then Some args.[2] else None

    let sArgs = Sprite.spriteArgs spriteName

    if not (File.Exists localPath) && not (Directory.Exists localPath) then
        eprintfn "Error: %s does not exist" localPath
        exit 1

    if Directory.Exists localPath then
        printfn "Pushing directory: %s -> %s" localPath remotePath

        let fullPath = Path.GetFullPath localPath
        let parent = Path.GetDirectoryName fullPath
        let baseName = Path.GetFileName fullPath

        // tar czf - -C parent baseName
        let tarPsi = ProcessStartInfo("tar")
        tarPsi.UseShellExecute <- false
        tarPsi.RedirectStandardOutput <- true

        for a in [| "czf"; "-"; "-C"; parent; baseName |] do
            tarPsi.ArgumentList.Add a

        use tarProc = Process.Start tarPsi

        // sprite exec [-s name] bash -c "mkdir -p DEST && tar xzf - -C DEST --strip-components=1"
        let spritePsi = ProcessStartInfo("sprite")
        spritePsi.UseShellExecute <- false
        spritePsi.RedirectStandardInput <- true

        for a in
            Array.append
                [| "exec" |]
                (Array.append sArgs
                    [| "bash"
                       "-c"
                       sprintf "mkdir -p '%s' && tar xzf - -C '%s' --strip-components=1" remotePath remotePath |]) do
            spritePsi.ArgumentList.Add a

        use spriteProc = Process.Start spritePsi

        tarProc.StandardOutput.BaseStream.CopyTo(spriteProc.StandardInput.BaseStream)
        spriteProc.StandardInput.Close()
        spriteProc.WaitForExit()
        tarProc.WaitForExit()

        if spriteProc.ExitCode <> 0 then
            eprintfn "Error: push failed"
            exit 1
    else
        printfn "Pushing file: %s -> %s" localPath remotePath

        let remoteDir = Path.GetDirectoryName remotePath

        let spritePsi = ProcessStartInfo("sprite")
        spritePsi.UseShellExecute <- false
        spritePsi.RedirectStandardInput <- true

        for a in
            Array.append
                [| "exec" |]
                (Array.append sArgs
                    [| "bash"
                       "-c"
                       sprintf "mkdir -p '%s' && cat > '%s'" remoteDir remotePath |]) do
            spritePsi.ArgumentList.Add a

        use spriteProc = Process.Start spritePsi
        use fileStream = File.OpenRead localPath
        fileStream.CopyTo(spriteProc.StandardInput.BaseStream)
        spriteProc.StandardInput.Close()
        spriteProc.WaitForExit()

        if spriteProc.ExitCode <> 0 then
            eprintfn "Error: push failed"
            exit 1

    printfn "Done."

File: ports/ocaml/lib/push.ml
(* push.ml — sprite-tool push: push a local file or directory to a sprite *)

let usage () =
  Printf.printf
{|Usage: sprite-tool push <local-path> <remote-path> [sprite-name]

Examples:
  sprite-tool push ./file.txt /home/sprite/file.txt
  sprite-tool push ./mydir /home/sprite/mydir
  sprite-tool push ./file.txt /tmp/file.txt ember-red-hawk
|};
  exit 1

let resolve_sprite_name args =
  match args with
  | _ :: _ :: name :: _ -> name
  | _ ->
    (match Sys.getenv_opt "SPRITE_NAME" with
     | Some name when String.length (String.trim name) > 0 -> String.trim name
     | _ ->
       Printf.eprintf "Error: sprite name not provided and SPRITE_NAME not set\n%!";
       exit 1)

let run args =
  match args with
  | [] | [_] -> usage ()
  | _ ->
    let local_path = List.nth args 0 in
    let remote_path = List.nth args 1 in
    let sprite_name = resolve_sprite_name args in
    let dry_run = false in

    if not (Sys.file_exists local_path) then begin
      Printf.eprintf "Error: %s does not exist\n%!" local_path;
      exit 1
    end;

    if Sys.is_directory local_path then begin
      Printf.printf "Pushing directory: %s -> %s\n%!" local_path remote_path;
      (* For directory push: tar up and extract with strip-components *)
      let parent_dir = Filename.dirname local_path in
      let base_name = Filename.basename local_path in
      let cmd =
        Printf.sprintf "tar czf - -C %s %s | sprite exec -s %s bash -c %s"
          (Filename.quote parent_dir)
          (Filename.quote base_name)
          (Filename.quote sprite_name)
          (Filename.quote (Printf.sprintf "mkdir -p '%s' && tar xzf - -C '%s' --strip-components=1"
            remote_path remote_path))
      in
      let code = Sprite.system_checked cmd in
      if code <> 0 then begin
        Printf.eprintf "Error: push directory failed (exit %d)\n%!" code;
        exit 1
      end
    end else begin
      Printf.printf "Pushing file: %s -> %s\n%!" local_path remote_path;
      let remote_dir = Filename.dirname remote_path in
      let _ = Sprite.sx ~dry_run ~sprite_name
        (Printf.sprintf "mkdir -p '%s'" remote_dir) in
      let cmd =
        Printf.sprintf "sprite exec -s %s bash -c %s < %s"
          (Filename.quote sprite_name)
          (Filename.quote (Printf.sprintf "cat > '%s'" remote_path))
          (Filename.quote local_path)
      in
      let code = Sprite.system_checked cmd in
      if code <> 0 then begin
        Printf.eprintf "Error: push file failed (exit %d)\n%!" code;
        exit 1
      end
    end;

    Printf.printf "Done.\n%!"
